src/Microsoft.ML.Transforms/Text/CharTokenizeTransform.cs

```diff
@@ -151,266 +140,316 @@ public override void Save(ModelSaveContext ctx)
-                // illegal codepoints, spaces, etc.) it is better to use something else
-                // to represent the character.
-                int[] boundaries = new int[CharsCount + 1];
-                var bldr = new StringBuilder();
-                for (int i = 1; i <= CharsCount; i++)
+                _parent = parent;
+                _types = new ColumnType[_parent.ColumnPairs.Length];
+                _isSourceVector = new bool[_parent.ColumnPairs.Length];
+                var keyType = new KeyType(DataKind.U2, 1, CharsCount);
+                for (int i = 0; i < _types.Length; i++)
                 {
-                    AppendCharRepr((char)i, bldr);
-                    boundaries[i] = bldr.Length;
+                    _isSourceVector[i] = inputSchema[_parent.ColumnPairs[i].input].Type.IsVector;
+                    _types[i] = new VectorType(keyType);
                 }
-
-                Host.Assert(bldr.Length == boundaries[boundaries.Length - 1]);
-                Interlocked.CompareExchange(ref _keyValuesBoundaries, boundaries, null);
-                Interlocked.CompareExchange(ref _keyValuesStr, bldr.ToString(), null);
-                bldr.Length = 0;
             }
 
-            var keyValuesStr = _keyValuesStr;
-            var keyValuesBoundaries = _keyValuesBoundaries;
-            Host.AssertValue(keyValuesBoundaries);
-
-            var values = dst.Values;
-            if (Utils.Size(values) < CharsCount)
-                values = new ReadOnlyMemory<char>[CharsCount];
-            for (int i = 0; i < CharsCount; i++)
-                values[i] = keyValuesStr.AsMemory().Slice(keyValuesBoundaries[i], keyValuesBoundaries[i + 1] - keyValuesBoundaries[i]);
-            dst = new VBuffer<ReadOnlyMemory<char>>(CharsCount, values, dst.Indices);
-        }
-
-        private void AppendCharRepr(char c, StringBuilder bldr)
-        {
-            // Special handling of characters identified in https://en.wikipedia.org/wiki/Unicode_control_characters,
-            // as well as space, using the control pictures.
-            if (c <= 0x20)
+            public override Schema.Column[] GetOutputColumns()
             {
-                // Use the control pictures unicode code block.
-                bldr.Append('<');
-                bldr.Append((char)(c + '\u2400'));
-                bldr.Append('>');
-                return;
+                var result = new Schema.Column[_parent.ColumnPairs.Length];
+                for (int i = 0; i < _parent.ColumnPairs.Length; i++)
+                {
+                    var builder = new Schema.Metadata.Builder();
+                    AddMetadata(i, builder);
+                    result[i] = new Schema.Column(_parent.ColumnPairs[i].output, _types[i], builder.GetMetadata());
+                }
+                return result;
             }
-            if ('\uD800' <= c && c <= '\uDFFF')
+
+            private void AddMetadata(int iinfo, Schema.Metadata.Builder builder)
```

Comment:
Minor nit pick here - since this is a helper function, I would make this a static function and pass in the metadata instead of using the InputSchema and _parent of the class. You probably do not need to pass in iinfo as this would be used to reference the metadata to this function call..